### DISCUSSION TITLE
Using Fine Tuning Models with GPT3Discord

### DISCUSSION THREAD

**Discussion by erinheit451** [Kav-K/GPTDiscord#125] [Q&A] [answered]
First, I just wanted to say thinks to the developers who've put so much time and love into this project. It's really cool and the care shows in all the details. 

I'm wanting to use the /converse function of the bot but switch out the davinci-003 model for a custom fine tune model. I'm a novice developer who jumped into projects once ChatGPT exploded in December and after working through replacing everything I can see in terms of model= variables that made sense in the openai model and text service files but I'm clearly missing something. 

What part of the application is responsible for defining the model used for the /conversation functionality of the bot? And is there any reason that I can't replace the davinci-003 model for my own fine tune? (Given all the pinecone DB calls and summarization stuff I didn't know if this idea was a nonstarter in terms of interdependencies). 

Thanks,
Erin

**Comment by cherryroots** [comment, marked as answer]
See [text_service_cog.py#L869](https://github.com/Kav-K/GPT3Discord/blob/a3ce3577a7d243357733a70989f412f16345b6e9/cogs/text_service_cog.py#L869)
Just set it to equal your custom model @erinheit451 

If you don't want to set one model for all conversation that's hardcoded you can add an entry to the Models class in openai_model.py then add that to the TEXT_MODELS list  then set your default model with the settings command, should use that then

```python
class Models:
    # Text models
    DAVINCI = "text-davinci-003"
    DAVINCI_FT = "some-model"
    CURIE = "text-curie-001"
    BABBAGE = "text-babbage-001"
    ADA = "text-ada-001"

    .....

    # Model collections
    TEXT_MODELS = [DAVINCI, DAVINCI_FT, CURIE, BABBAGE, ADA, CODE_DAVINCI, CODE_CUSHMAN]
    EDIT_MODELS = [EDIT, CODE_EDIT]

    ....

    # Tokens Mapping
    TOKEN_MAPPING = {
        "text-davinci-003": 4024,
        "some-model": 4024,
        "text-curie-001": 2024,
        "text-babbage-001": 2024,
        "text-ada-001": 2024,
        "code-davinci-002": 7900,
        "code-cushman-001": 2024,
    }
```

**Reply by erinheit451** [reply]
Is there any other dependency that would keep it from updating the parameters list in Discord to make the model selectable after the code has been updated?  I included my code for reference in case I've overlooked something. I tried a TEST model in case the length or the colon in my file name would matter. I also tried just flat hardcoding my model for DAVINCI = to see, and that didn't make a difference either which seems bizarre.  But the Discord menu happily updates for things like new opener text files when I upload those so it is taking updates to the application. 

Prior to posting this question last night I went on a rabbit hunt just looking for all the places the code references the model or defines it and tried making them hard-coded and it still didn't seem to change it. Thanks again for the help and such a prompt reply. Maybe I'm just missing something fundamental. 

class Models:
    # Text models
    DAVINCI = "text-davinci-003"
    EMBER = "davinci:ft-personal-2023-01-29-22-37-00"
    CURIE = "text-curie-001"
    BABBAGE = "text-babbage-001"
    ADA = "text-ada-001"
    TEST = "text-test-001"
---
    # Model collections
    TEXT_MODELS = [DAVINCI, EMBER, CURIE, BABBAGE, ADA, CODE_DAVINCI, CODE_CUSHMAN, TEST]
    EDIT_MODELS = [EDIT, CODE_EDIT]
---
    # Tokens Mapping
    TOKEN_MAPPING = {
        "text-davinci-003": 4024,
        "davinci:ft-personal-2023-01-29-22-37-00": 4024, 
        "text-curie-001": 2024,
        "text-babbage-001": 2024,
        "text-ada-001": 2024,
        "code-davinci-002": 7900,
        "code-cushman-001": 2024,
        "text-test-001": 4024,
    }
![Thread](https://user-images.githubusercontent.com/18537530/215651828-c4a5aaf2-2f1a-4e21-90c4-4d067fda6a5b.jpeg)
![values](https://user-images.githubusercontent.com/18537530/215651835-c1321934-463d-4ec1-89f1-ebc7ac188d62.jpg)

**Reply by cherryroots** [reply]
Should change after that, the autocomplete for the model value just gets the stuff in the `TEXT_MODELS` array. Done on main with no other changes
![image](https://user-images.githubusercontent.com/15952522/215730030-89f5571e-ae45-437b-a311-501f56e33e8f.png)
![image](https://user-images.githubusercontent.com/15952522/215730064-f0b30263-9353-46f3-ab56-22a56ba7542b.png)

**Comment by cherryroots** [comment]
I'll see about finally adding the custom model option to the conversations today when running the converse command

**Reply by erinheit451** [reply]
That would be amazing! Thank you so much. <3
